presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -77,11 +81,23 @@
     private final boolean recursiveDirWalkerEnabled;
     private final Executor executor;
     private final ConnectorSession session;
-    private final AtomicInteger outstandingTasks = new AtomicInteger();
     private final ConcurrentLazyQueue<HivePartitionMetadata> partitions;
     private final Deque<HiveFileIterator> fileIterators = new ConcurrentLinkedDeque<>();
     private final AtomicInteger remainingInitialSplits;
 
+    // Purpose of this lock:
+    // * When write lock is acquired, except the holder, no one can do any of the following:
+    // ** poll from partitions
+    // ** poll from or push to fileIterators
+    // ** push to hiveSplitSource
+    // * When any of the above three operations is carried out, either a read lock or a write lock must be held.
+    // * When a series of operations involving two or more of the above three operations are carried out, the lock
+    //   must be continuously held throughout the series of operations.
+    // Implications:
+    // * if you hold a read lock but not a write lock, you can do any of the above three operations, but you may
+    //   see a series of operations involving two or more of the operations carried out half way.
+    private final ReentrantReadWriteLock threeQueuesRwLock = new ReentrantReadWriteLock();
```

Comment:
I see that there are only 2 classes named queue. And therefore `threeQueues` is a little confusing. But how does `queues` solve that? Doesn't it make things worse because reader would just assume that this lock only cover the two classes named queue?